File: 2 - Sorting and Order Statistics/2.4 - Medians and Order Statistics/Code/Select.js
/**
 * Select.js
 *
 */

class Select {
    constructor(array) {
        this.array = array;
    }

    partitionAround(p, r, x) {
        let i = p - 1;

        for (let j = p; j < r; j++) {
            if (this.array[j] <= x) {
                i += 1;
                let temp = this.array[i];
                this.array[i] = this.array[j];
                this.array[j] = temp;
            }
        }

        let temp = this.array[i + 1];
        this.array[i + 1] = this.array[r];
        this.array[r] = temp;

        return i + 1;
    }

    select(p, r, i) {
        while ((r - p + 1) % 5 != 0) {
            for (let j = p + 1; j < r + 1; j++) {
                if (this.array[p] > this.array[j]) {
                    let temp = this.array[j];
                    this.array[j] = this.array[p];
                    this.array[p] = temp;
                }
            }

            if (i == 1) {
                return this.array[p];
            }

            p += 1;
            i -= 1;
        }

        let g = Math.floor((r - p + 1) / 5);

        for (let j = p; j < p + g; j++) {
            let indices = [j, j + g, j + 2 * g, j + 3 * g, j + 4 * g].filter(
                (index) => index <= r
            );
            let elements = indices.map((index) => this.array[index]);

            elements.sort((a, b) => a - b);

            for (let k = 0; k < indices.length; k++) {
                let index = indices[k];
                let sortedValue = elements[k];
                this.array[index] = sortedValue;
            }
        }

        let x = this.select(p + 2 * g, p + 3 + g - 1, Math.ceil(g / 2));
        let q = this.partitionAround(p, r, x);

        let k = q - p + 1;

        if (i == k) {
            return this.array[q];
        }

        if (i < k) {
            return this.select(p, q - 1, i);
        }

        return this.select(q + 1, r, i - k);
    }
}

module.exports = Select;

File: 2 - Sorting and Order Statistics/2.4 - Medians and Order Statistics/Code/select.py
"""
select.py

"""


from math import floor, ceil


class Select:
    """Implements Select Algorithm"""

    def __init__(self, array):
        self.array = array

    def partiton_around(self, p, r, x):
        """Modified partition method, the same as the one used in 'randomized_select', only the
        pivot element is passed in as an argument.

        Returns:
            int: New pivot element index
        """
        i = p - 1

        for j in range(p, r + 1):
            if self.array[j] <= x:
                i += 1
                self.array[i], self.array[j] = self.array[j], self.array[i]

        self.array[i + 1], self.array[r] = self.array[r], self.array[i + 1]

        return i + 1

    def select(self, p, r, i):
        """A modified version of 'randomized_select' with the same arguments

        Time complexity: O(n)
        Space complexity: O(log n) - recursion depth

        Returns:
            int: Value of the i-th order statistic
        """
        while (r - p + 1) % 5 != 0:
            for j in range(p + 1, r + 1):
                if self.array[p] > self.array[j]:
                    self.array[p], self.array[j] = self.array[j], self.array[p]

            if i == 1:
                return self.array[p]

            p += 1
            i -= 1

        g = floor((r - p + 1) / 5)

        for j in range(p, p + g - 1):
            indices = [j, j + g, j + 2 * g, j + 3 * g, j + 4 * g]
            elements = [self.array[index] for index in indices]
            elements.sort()

            for index, sorted_value in zip(indices, elements):
                self.array[index] = sorted_value

        x = self.select(p + 2 * g, p + 3 + g - 1, ceil(g / 2))
        q = self.partiton_around(p, r, x)

        k = q - p + 1

        if i == k:
            return self.array[q]

        if i < k:
            return self.select(p, q - 1, i)

        return self.select(q + 1, r, i - k)
